File: v1.3/GPU/source/image.cpp
#define IMAGE_CC

#include "utils.h"
#include "image.h"

real _xcell, _ycell, _zcell;
real _ixcell, _iycell, _izcell;
real eps_cell;
real beta_sin, beta_cos, beta_term, gamma_sin, gamma_cos, gamma_term;
real _box34;
int monoclinic=0,triclinic=0,octahedron=0;
BoxShape_e boxShape;
//extern const int tinkerdebug;

long long gcd(long long a, long long b)
{
   while (b != 0) {
      long long t = b;
      b = a % b;
      a = t;
   }
   return llabs(a);
}

EXTERN_C_BEG

void C_get_cell( real xcell_, real ycell_, real zcell_, real eps_cell_, real beta_sin_, real beta_cos_,
                 real beta_term_, real gamma_sin_, real gamma_cos_, real gamma_term_,
                 int mono_, int tric_, int octa_, real box34_ ){

   assert( mono_+tric_+octa_<2 );

   _xcell     = xcell_;
   _ycell     = ycell_;
   _zcell     = zcell_;
   _ixcell    = (real) CodePrm::onem/ (double)xcell_;
   _iycell    = (real) CodePrm::onem/ (double)ycell_;
   _izcell    = (real) CodePrm::onem/ (double)zcell_;
   eps_cell   = eps_cell_;
   beta_sin   = beta_sin_;
   beta_cos   = beta_cos_;
   beta_term  = beta_term_;
   gamma_sin  = gamma_sin_;
   gamma_cos  = gamma_cos_;
   gamma_term = gamma_term_;
   monoclinic = mono_;
   triclinic  = tric_;
   octahedron = octa_;
   _box34     = box34_;

                boxShape = BOX_ORTH;
   if ( mono_ ) boxShape = BOX_MONO;
   if ( tric_ ) boxShape = BOX_TRIC;
   if ( octa_ ) boxShape = BOX_OCTA;

   //if (tinkerdebug) {
   //   printf(" C_get_cell -beta %lf %lf -gamma %lf %lf -shape %d %d %d\n", beta_sin,beta_term,gamma_sin,gamma_term,monoclinic,triclinic,octahedron,_box34);
   //}

}

void best_rational_approx(const double x, const long long max_den,
                          long long *num, long long *den)
{
    long long h0 = 0, h1 = 1;
    long long k0 = 1, k1 = 0;

    double value = x;
    int sign = (x < 0) ? -1 : 1;
    value = fabs(value);

    while (1)
    {
        long long a = (long long)floor(value);

        long long h2 = a * h1 + h0;
        long long k2 = a * k1 + k0;

        if (k2 > max_den)
           break;

        h0 = h1; h1 = h2;
        k0 = k1; k1 = k2;

        double frac = value - a;
        if (frac < 1e-16)
            break;

        value = 1.0 / frac;
    }

    *num = sign * h1;
    *den = k1;

    long long g = gcd(*num, *den);
    *num /= g;
    *den /= g;
    //printf("  -best_rational_approx %24.15f -m_den %lu = ( %lu / %lu )\n",x,max_den,*num,*den);
}

EXTERN_C_END

File: v1.3/GPU/source/image.f90
!
!     Sorbonne University
!     Washington University in Saint Louis
!     University of Texas at Austin
!
!     ################################################################
!     ##                                                            ##
!     ##  subroutine image  --  compute the minimum image distance  ##
!     ##                                                            ##
!     ################################################################
!
!
!     "image" takes the components of pairwise distance between
!     two points in a periodic box and converts to the components
!     of the minimum image distance
!
!
#include "tinker_precision.h"
subroutine image (xr,yr,zr)
   use sizes
   use boxes
   use cell
   implicit none
   real(t_p) xr,yr,zr,cel
!
!
!     for orthogonal lattice, find the desired image directly
!
   if (orthogonal) then
      if (abs(xr) .gt. xcell2) then
         cel = sign(xcell,xr)
         xr  = xr - cel*floor((abs(xr)+xcell2)/xcell)
      end if
      if (abs(yr) .gt. ycell2) then
         cel = sign(ycell,yr)
         yr  = yr - cel*floor((abs(yr)+ycell2)/ycell)
      end if
      if (abs(zr) .gt. zcell2) then
         cel = sign(zcell,zr)
         zr  = zr - cel*floor((abs(zr)+zcell2)/zcell)
      end if
!
!     for monoclinic lattice, convert "xr" and "zr" to
!     fractional coordinates, find desired image and then
!     translate fractional coordinates back to Cartesian
!
   else if (monoclinic) then
      zr = zr / beta_sin
      xr = xr - zr*beta_cos
      do while (abs(xr) .gt. xcell2)
         xr = xr - sign(xcell,xr)
      end do
      do while (abs(yr) .gt. ycell2)
         yr = yr - sign(ycell,yr)
      end do
      do while (abs(zr) .gt. zcell2)
         zr = zr - sign(zcell,zr)
      end do
      xr = xr + zr*beta_cos
      zr = zr * beta_sin
!
!     for triclinic lattice, convert pairwise components to
!     fractional coordinates, find desired image and then
!     translate fractional coordinates back to Cartesian
!
   else if (triclinic) then
      zr = zr / gamma_term
      yr = (yr - zr*beta_term) / gamma_sin
      xr = xr - yr*gamma_cos - zr*beta_cos
      do while (abs(xr) .gt. xcell2)
         xr = xr - sign(xcell,xr)
      end do
      do while (abs(yr) .gt. ycell2)
         yr = yr - sign(ycell,yr)
      end do
      do while (abs(zr) .gt. zcell2)
         zr = zr - sign(zcell,zr)
      end do
      xr = xr + yr*gamma_cos + zr*beta_cos
      yr = yr*gamma_sin + zr*beta_term
      zr = zr * gamma_term
!
!     for truncated octahedron, use orthogonal box equations,
!     then perform extra tests to remove corner pieces
!
   else if (octahedron) then
      do while (abs(xr) .gt. xcell2)
         xr = xr - sign(xcell,xr)
      end do
      do while (abs(yr) .gt. ycell2)
         yr = yr - sign(ycell,yr)
      end do
      do while (abs(zr) .gt. zcell2)
         zr = zr - sign(zcell,zr)
      end do
      if (abs(xr)+abs(yr)+abs(zr) .gt. box34) then
         xr = xr - sign(xcell2,xr)
         yr = yr - sign(ycell2,yr)
         zr = zr - sign(zcell2,zr)
      end if
   end if
   return
end
!
subroutine image_acc(xr,yr,zr)
!$acc routine seq
   use boxes,only: orthogonal,octahedron&
      &,box34
   use cell
   use tinheader ,only: ti_p
   implicit none
   real(t_p) xr,yr,zr
   real(t_p) cel

   if (orthogonal) then
      if (abs(xr) .gt. xcell2) then
         cel = sign(xcell,xr)
         xr  = xr - cel*floor((abs(xr)+xcell2)/xcell)
      end if
      if (abs(yr) .gt. ycell2) then
         cel = sign(ycell,yr)
         yr  = yr - cel*floor((abs(yr)+ycell2)/ycell)
      end if
      if (abs(zr) .gt. zcell2) then
         cel = sign(zcell,zr)
         zr  = zr - cel*floor((abs(zr)+zcell2)/zcell)
      end if
   else if (octahedron) then
      if (abs(xr) .gt. xcell2)&
         &xr  = xr - sign(xcell,xr)&
         &*floor((abs(xr)-xcell2)*i_xcell + 1.0_ti_p)
      if (abs(yr) .gt. ycell2)&
         &yr  = yr - sign(ycell,yr)&
         &*floor((abs(yr)-ycell2)*i_ycell + 1.0_ti_p)
      if (abs(zr) .gt. zcell2)&
         &zr  = zr - sign(zcell,zr)&
         &*floor((abs(zr)-zcell2)*i_zcell + 1.0_ti_p)
      if (abs(xr)+abs(yr)+abs(zr) .gt. box34) then
         xr = xr - sign(xcell2,xr)
         yr = yr - sign(ycell2,yr)
         zr = zr - sign(zcell2,zr)
      end if
   end if
end
!
subroutine image1d(val,vcell,vcell2)
!$acc routine seq
   use tinheader,only:ti_p
   implicit none
   real(ti_p),intent(inout):: val
   real(ti_p),intent(in) :: vcell,vcell2

   if (abs(val) .gt. vcell2)&
      &val  = val - sign(vcell,val)*floor((abs(val)+vcell2)/vcell)
end subroutine
!
!    subroutine imagefrac: takes a vector in fractional coordinate and
!    get the associated minimum image convention vector in fractional coordinate
!

subroutine imagefrac(ar,br,cr)
   use boxes
   use cell
   implicit none
   real(t_p) ar,br,cr, cel
   if (abs(ar) .gt. xcell2) &
      ar = ar - sign(xcell,ar)*floor((abs(ar)+xcell2)/xcell)
   if (abs(br) .gt. ycell2) &
      br = br - sign(ycell,br)*floor((abs(br)+ycell2)/ycell)
   if (abs(cr) .gt. zcell2) &
      cr = cr - sign(zcell,cr)*floor((abs(cr)+zcell2)/zcell)
end
!
!    subroutine imagecell: takes a vector in "cell" coordinate and
!    get the associated minimum image convention vector in "cell" coordinate
!

subroutine imagecell(n_ar,n_br,n_cr,na,nb,nc)
   use boxes
   use cell
   implicit none
   integer n_ar,n_br,n_cr,na,nb,nc
   do while (abs(n_ar) .gt. int(na/2))
      n_ar = n_ar - sign(na,n_ar)
   end do
   do while (abs(n_br) .gt. int(nb/2))
      n_br = n_br - sign(nb,n_br)
   end do
   do while (abs(n_cr) .gt. int(nc/2))
      n_cr = n_cr - sign(nc,n_cr)
   end do
end
!
!     subroutine ctfvec: takes a vector in cartesian coordinates and put
!     it in fractional coordinates
!
subroutine ctfvec(xr,yr,zr,ar,br,cr)
use boxes
   implicit none
   real(t_p) xr,yr,zr
   real(t_p) ar,br,cr
   ar = ctfmat(1,1)*xr + ctfmat(1,2)*yr + ctfmat(1,3)*zr
   ar = xbox*ar
   br = ctfmat(2,1)*xr + ctfmat(2,2)*yr + ctfmat(2,3)*zr
   br = ybox*br
   cr = ctfmat(3,1)*xr + ctfmat(3,2)*yr + ctfmat(3,3)*zr
   cr = zbox*cr
end

!subroutine ftcvec!: takes a vector in fractional coordinates and put it in cartesian coordinates

subroutine ftcvec(ar,br,cr,xr,yr,zr)
   use boxes
   implicit none
   real(t_p) xr,yr,zr
   real(t_p) ar,br,cr
   real(t_p) artemp,brtemp,crtemp
   artemp = ar/xbox
   brtemp = br/ybox
   crtemp = cr/zbox
   xr = ftcmat(1,1)*artemp + ftcmat(2,1)*brtemp + ftcmat(3,1)*crtemp
   yr = ftcmat(1,2)*artemp + ftcmat(2,2)*brtemp + ftcmat(3,2)*crtemp
   zr = ftcmat(1,3)*artemp + ftcmat(2,3)*brtemp + ftcmat(3,3)*crtemp
end
